add file transfer docs (#1930)
emqtt/emq-docs-cn: en_US/file-transfer/broker.md

> # Configure File Transfer Setting in EMQX

Reviewer: Might be just
```suggestion
# Configuration
```
then?

This page provides instructions on configuring the file transfer settings in EMQX. Topics will cover the MQTT settings for file transfer operations, as well as the storage options available for managing file metadata, segments, and exporting files. Specifically, instructions will focus on the local storage backend and the S3 exporter for file export.

EMQX allows not to configure most of the file transfer settings and use the default values.

## Configure File Transfer MQTT Settings

To optimize file transfer operations and prevent clients from waiting excessively, we can set specific timeouts for different file transfer operations. The following MQTT settings can be configured:
```
file_transfer {
    enable = true
    init_timeout = "10s"
    store_segment_timeout = "10s"

>     assemble_timeout = "60s"

Reviewer: IMO these are rarely needed to be configured, I'd move mentions of them at the very end of the chapter.

Reviewer 2: Moved this part of configurations to the end.

}
```

- `init_timeout`: Timeout for initialization operation.
- `store_segment_timeout`: Timeout for storing file segments.
- `assemble_timeout`: Timeout for file assembly.

If any of these operations exceed the specified timeout, the MQTT client will receive a PUBACK packet with the `RC_UNSPECIFIED_ERROR` code.

## Configure File Transfer Storage

EMQX provides storage options for managing file metadata, segments, and exporting files. Currently, the EMQX supports a single storage backend: local file storage.

To enable local file storage, use the following configuration:

```
{
    file_transfer {
        enable = true
        storage {
            local {
                enable = true
            }
        }
    }
}
```

With the local file storage, EMQX stores file metadata and segments in the local file system of the receiving nodes. Additionally, the local file storage exports uploaded files using a configured exporter. EMQX supports two exporters: the local exporter and the S3 exporter, exporting files correspondingly to the local file system and S3-compatible object storage.

> Several exporters may have settings simultaneously, but only one may be enabled.

Reviewer: Looks a bit out of place right now. Would probably look better below, after _Segment settings_, maybe in a separate `### Exporter settings`?

Reviewer 2: I agree that it looks weird to have a single sentence in a separated paragraph, but I think it does belong to the general descriptions about the storage configuration. So, I moved it up.

- Segment garbage collection settings.

```
{
    file_transfer {
        enable = true
        storage {
            local {
                enable = true
                segments {
                    root = "/var/lib/emqx/file_transfer/segments"
                    gc {
                        interval = "1h"
                        maximum_segments_ttl = "1d"
                        minimum_segments_ttl = "1h"
                    }
                }
            }
        }
    }
}
```

It is important to choose appropriate values for these parameters based on factors such as expected file transfer load, file sizes, concurrent transfers, and available disk space.

### Local Exporter Settings

The local exporter allows exporting files to a local file system. It only has the setting of the root folder for the exported files.

```
{
    file_transfer {
        enable = true
        storage {
            local {
                enable = true
                exporter {
                    local {
                        enable = true
                        root = "/var/lib/emqx/file_transfer/exported"
                    }
                }
            }
        }
    }
}
```

> Exported files may accumulate over time, potentially resulting in a large number of files in the export folder. To address this, EMQX utilizes a bucketing scheme for storing exported files. The scheme involves the following 6-level folder hierarchy:

Reviewer: Nit.
```suggestion
Exported files may accumulate over time, potentially resulting in a large number of files in the export folder. To address this, EMQX utilizes a bucketing scheme to store exported files. The scheme involves the following 6-level folder hierarchy:
```

> - It calculates the sha256 hash of the file id together client id (`ABCDEFG012345...`).

Reviewer: ```suggestion
- Calculate the hash of the file id together with client id (`ABCDEFG012345...`).
```

> - It stores the file into a 6-level folder hierarchy, using as levels:

Reviewer: Nit.
```suggestion
- Store the file in a 6-level directory hierarchy, with each level defined as follows:
```
^ Again, isn't _directory_ closer than Windows-slang _folder_ to what our users are accustomed to.

  * The first 2 bytes of the hash as the first-level folder name;
  * The next 2 bytes as the second-level folder name;
  * The rest of the hash as the 3-rd level folder name;
  * The client id;
  * The file id;

>   * The escaped file name as the last level.

Reviewer: Isn't the other way around? Client / File IDs are escaped, filename is not.

Tangentially, I wonder if we really need to describe these implementation details here.

Reviewer 3: It is possible to inform the user as an advanced option to manage these "file fragments".

Reviewer: Sure, got it. Point about escaping still stands though.

Author: ✅
